mmv1/third_party/terraform/services/composer/resource_composer_environment.go.erb

```diff
@@ -682,6 +682,36 @@ func ResourceComposerEnvironment() *schema.Resource {
 								},
 							},
 						},
+
+						"data_retention_config": {
+							Type:         schema.TypeList,
+							Optional:     true,
+							Computed:     true,
+							AtLeastOneOf: composerConfigKeys,
+							MaxItems:     1,
+							Description:  `The configuration setting for Airflow data retention mechanism. This field is supported for Cloud Composer environments in versions composer-2.0.32-airflow-2.1.4. or newer`,
+							Elem: &schema.Resource{
+								Schema: map[string]*schema.Schema{
+									"task_logs_retention_config": {
+										Type:         schema.TypeList,
+										Optional:     true,
+										ForceNew:     false,
+										Description: `Optional. The configuration setting for Task Logs.`,
+										Elem: &schema.Resource{
+											Schema: map[string]*schema.Schema{
+												"storage_mode": {
+													Type:         schema.TypeString,
+													Optional:     true,
+													ForceNew:     false,
+													ValidateFunc: validation.StringInSlice([]string{"CLOUD_LOGGING_ONLY", "CLOUD_LOGGING_AND_CLOUD_STORAGE"}, false),
```

Comment:
If you are expecting this list to expand regularly, it could make sense to leave off validation, since that allows versions of the provider to be a little bit future-compatible.

Reply:
It's quite improbable to have frequent changes to this list.